### PR TITLE
break up the huge fsm.go file, 

### PR DESCRIPTION
**Pull request by sclasen**
introduce the FSMSerializer interface to break the circular dep between FSM and FSMContext

### CODE REVIEW

File: fsm_models.go [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+	// reflection will asplode if we try to use nil
+	if data == nil {
+		data = reflect.New(reflect.TypeOf(f.stateData)).Interface()
+	}
+	return m.v.Call([]reflect.Value{reflect.ValueOf(f), reflect.ValueOf(h), reflect.ValueOf(data)})[0].Interface().(Outcome)
+}
+
+// FSMSerializer is the contract for de/serializing state inside an FSM, typically implemented by the FSM itself
+// but serves to break the circular dep between FSMContext and FSM.
+type FSMSerializer interface{
```
Comment (sclasen) [inline fsm_models.go]: This is the introduced interface to break the circular dep.

File: fsm_models.go [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+	// reflection will asplode if we try to use nil
+	if data == nil {
+		data = reflect.New(reflect.TypeOf(f.stateData)).Interface()
+	}
+	return m.v.Call([]reflect.Value{reflect.ValueOf(f), reflect.ValueOf(h), reflect.ValueOf(data)})[0].Interface().(Outcome)
+}
+
+// FSMSerializer is the contract for de/serializing state inside an FSM, typically implemented by the FSM itself
+// but serves to break the circular dep between FSMContext and FSM.
+type FSMSerializer interface {
```
Comment (sclasen) [inline fsm_models.go]: This is the introduced interface to break the circular dep btw FSM and FSMContext

**Comment by sclasen** [comment]
@fabiokung any thoughts on this? mostly just a cut and paste of code except for the FSMSerializer interface.  -control compiles/tests with no modification so  probably all good.

**Comment by sclasen** [comment]
bringin in

**Comment by fabiokung** [comment]
:+1: sorry for the late comment. At a first glance, it feels a bit weird that `FSMSerializer` has a `StateSerializer` (usually `JSONStateSerializer`) inside it. Maybe it's just naming, but nothing that can't be solved in later iterations.

**Comment by sclasen** [comment]
Thanks! Agree on the name being weird, hopefully just temporary;)

**Pull request merged**